src/eve/pattern_matching.py

```diff
@@ -0,0 +1,139 @@
+            for diff in get_differences(el_a, el_b, path=f"{path}[{i}]"):
+                yield diff
+
+
+def _get_differences_dict(a: dict, b: Any, path: str = "") -> Iterator[tuple[str, str]]:
+    if not isinstance(b, dict):
+        yield (path, f"Expected dict, but got {type(b).__name__}")
+    elif set(a.keys()) != set(b.keys()):
+        a_min_b = set(a.keys()).difference(b.keys())
+        b_min_a = set(b.keys()).difference(a.keys())
+        if a_min_b:
+            missing_keys_str = "`" + "`, `".join(map(str, a_min_b)) + "`"
+            yield (
+                path,
+                f"Expected dictionary with keys `{'`, `'.join(map(str, a.keys()))}`, but the following keys are missing: {missing_keys_str}",
+            )
+        if b_min_a:
+            extra_keys_str = "`" + "`, `".join(map(str, b_min_a)) + "`"
+            yield (
+                path,
+                f"Expected dictionary with keys `{'`, `'.join(map(str, a.keys()))}`, but the following keys are extra: {extra_keys_str}",
+            )
+    else:
+        for k, v_a, v_b in zip(a.keys(), a.values(), b.values()):
+            yield from get_differences(v_a, v_b, path=f'{path}["{k}"]')
+
+
+def get_differences(a: Any, b: Any, path: str = "") -> Iterator[tuple[str, str]]:
+    """Compare two objects and return a list of differences.
+
+    If the arguments are lists or dictionaries comparison is recursively per item. Objects are compared
+    using equality operator or if the left-hand-side is an `ObjectPattern` its type and attributes
+    are compared to the right-hand-side object. Only the attributes of the `ObjectPattern` are used
+    for comparison, disregarding potential additional attributes of the right-hand-side.
+    """
+    if isinstance(a, ObjectPattern):
+        yield from _get_differences_object_pattern(a, b, path)
+    elif isinstance(a, list):
+        yield from _get_differences_list(a, b, path)
+    elif isinstance(a, dict):
+        yield from _get_differences_dict(a, b, path)
+    else:
+        if type(a) != type(b):
+            yield (path, f"Expected a value of type {type(a).__name__}, but got {type(b).__name__}")
+        elif a != b:
+            yield (path, f"Values are not equal. `{a}` != `{b}`")
+
+
+@dataclass(frozen=True)
+class ModuleWrapper:
+    """
+    Small wrapper to conveniently create `ObjectPattern`s for classes of a module.
+
+    Example:
+    >>> import foo_ir  # doctest: +SKIP
+    >>> foo_ir_ = ModuleWrapper(foo_ir)  # doctest: +SKIP
+    >>> assert foo_ir_.Foo(bar="baz").matches(foo_ir.Foo(bar="baz", foo="bar"))  # doctest: +SKIP
+    >>> assert not foo_ir_.Foo(bar="bar").matches(foo_ir.Foo(bar="baz", foo="bar"))  # doctest: +SKIP
+    """
+
+    module: ModuleType
+
+    def __getattr__(self, item: str) -> ObjectPatternConstructor:
+        val = getattr(self.module, item)
+        if not inspect.isclass(val):
+            raise ValueError("Only classes allowed.")
+        return ObjectPatternConstructor(val)
```

Comment:
I find this interface confusing and error-prone. I would rather use a more explicit pattern creator syntax, for example using __class_getitem__ to get something like `Pattern[SomeClass](foo="foo", bar="bar")` or `Matcher[SomeClass](foo="foo", bar="bar")`